File: dev/requirements/logic/cases/tap-live-card.5.4.case.swift
import Foundation
import SetlistCore

/// 5.4 — Tapping the live card changes nothing and opens no zero-length
/// segment. Mid-joke, a comedian's thumb finds its own card constantly.
let tapLiveCard_5_4 = RequirementCase(id: "5.4") {
    let airport = Joke(title: "Airport security", body: "…")
    var stage = StageModel(jokes: [airport], plannedLength: 1200)

    stage.tapCard(jokeID: airport.id, at: CaptureTime(0))
    let before = stage.segments
    let outcome = stage.tapCard(jokeID: airport.id, at: CaptureTime(84))

    try expectEqual(outcome, .ignored, "tap outcome")
    try expectEqual(stage.segments, before, "no segment was opened or closed")
    try expectEqual(stage.liveJokeID, airport.id, "the joke is still live")
    try expect(!stage.segments.contains(where: { $0.duration == 0 }), "no zero-length segment exists")
}

File: dev/requirements/logic/cases/tap-live-card.5.4.case.dart
import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 5.4 — Tapping the live card changes nothing and opens no zero-length segment.
final tapLiveCard_5_4 = RequirementCase('5.4', (tester) {
  final stage = StageModel(
    jokes: [
      Joke(id: 'opener', title: 'Opener', body: ''),
      Joke(id: 'van', title: "Dad's van", body: ''),
    ],
    plannedLength: const Duration(minutes: 20),
  );

  stage.tapCard(jokeId: 'opener', at: const CaptureTime(0));

  final outcome = stage.tapCard(jokeId: 'opener', at: const CaptureTime(30));

  expectThat(outcome, const TapIgnored(), 'the tap is ignored');
  expectThat(stage.segments.length, 1, 'no second segment is opened');
  expectAbsent(stage.segments.single.end, 'and the one that is open stays open');
  expectThat(stage.liveJokeId, 'opener', 'the same joke is still live');
});
